include/jemalloc/internal/extent.h

```diff
@@ -30,22 +30,29 @@ void ecache_dalloc(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
 edata_t *ecache_evict(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
     ecache_t *ecache, size_t npages_min);
 
+void extent_gdump_add(tsdn_t *tsdn, const edata_t *edata);
+void extent_record(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, ecache_t *ecache,
+    edata_t *edata);
 void extent_dalloc_gap(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
     edata_t *edata);
+edata_t *extent_alloc_wrapper(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
+    void *new_addr, size_t size, size_t alignment, bool zero, bool *commit,
+    bool growing_retained);
 void extent_dalloc_wrapper(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
     edata_t *edata);
 void extent_destroy_wrapper(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks,
     edata_t *edata);
 bool extent_commit_wrapper(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata,
-    size_t offset, size_t length);
+    size_t offset, size_t length, bool growing_retained);
 bool extent_decommit_wrapper(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata,
     size_t offset, size_t length);
 bool extent_purge_lazy_wrapper(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata,
     size_t offset, size_t length);
 bool extent_purge_forced_wrapper(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata,
     size_t offset, size_t length);
 edata_t *extent_split_wrapper(tsdn_t *tsdn, pac_t *pac,
-    ehooks_t *ehooks, edata_t *edata, size_t size_a, size_t size_b);
+    ehooks_t *ehooks, edata_t *edata, size_t size_a, size_t size_b,
+    bool holding_core_locks);
```

Comment:
One thing makes worry is the newly exposed details in the interfaces here, i.e. growing_retained, holding_core_locks and gdump_add which were previously all internal to extent.c.  I do like that the wrapper functions did not require any of these internal details, and it did take quite some refactoring effort to achieve that.

In a way it says that the san_bump.c is very related to extent.c. I was hoping adding the new `_commit_zero()` facility would help avoid some of the exposure here. But it looks that wasn't enough. Perhaps we should consider merging them? WDYT @Lapenkov?

Reply:
On the one hand, I don't like it exposed either, on the other hand the role of this argument is to simply ensure correct locking. I don't think it's a deal breaker to expose this sanity checking mechanism to the outside. I can make `locked` and `unlocked` versions of most of these functions, which would essentially achieve the same goal, without exposing `growing_retained` or `holding_core_locks`.

The only exception is `extent_alloc_wrapper`, where `gdump_add` depens on `growing_retained`. This relation is pretty simple if you think about it - when you allocate a retained extent, you don't add retained memory to gdump. When you allocate an extent as a mapping solely dedicated to it, you instantly add that mapping's size to gdump.

I don't like merging `extent` and `san_bump_alloc`, because `san_bump_alloc` is a stateful entity that uses `extent.h` functionality to manage extents. `extent` is a collection of functions to operate extents. I see clear distinction here, and with `extent.c` already being 1300 lines, feels like merging them will complicate things. I would rather put extra effort into exposing a more user-friendly interface through `extent.h` than merging it with `san_bump`.

Reply:
I'm convinced we should keep it separate. But still feel we should try tightening the interface changes. The biggest issue I have is the gdump_add(); usually that's done when registering the extent. It looks we register the whole extent w/o gdump_add when calling alloc_wrapper -- where do we register the new extent after the split?

A couple of other thoughts:
1) for alloc_wrapper(), will it work better if we pass in an additional `size_t over_reserve`, instead of the `bool growing_retained`? In other words, it's like having a new `alloc_grow_wrapper()` which handles the alloc + reserving together.
2) it looks `extent_commit_wrapper` doesn't have to be changed? Probably because we have the new commit_zero() now.
3) the new use of extent_record() if always in the failure case of commit_zero(), should we integrate extent_record into the failure path of commit_zero()?

Appreciate the effort and patience on tidying everything. We are getting very close now.

Reply:
Correct, we first register the whole extent allocated by `alloc_wrapper`, but don't add it to `gdump`. Later we cut a piece of that extent and, if no errors have occurred, add it to gdump (san_bump.c:88). `extent_grow_retained` works similarly.

1. Can you elaborate? By alloc + reserving you mean in should allocate a retained extent and cut a smaller piece of it?
2. Yes, reverting it back. Thanks.
3. I don't think so, if we add error handling there, it will be easy to miss it when reading the call site, unless we somehow reflect it in the function's name. And this function's name isn't ideal already :)

Reply:
Discussed offline and decided to evaluate if gdump management could be moved into `commit_zero` (and renaming the function).

I've checked, and it doesn't really fit in the picture. `commit_zero` is called in `extent_recycle`, which can recycle an extent from any ecache, not just retained. Hence, gdump management is done separately. So, either I do not reuse `commit_zero`, and add gdump management there, or I keep the code as is. To me, keeping as is seems better.